# The Movies Dataset

There are data from various sources to be merged into one dataset, with 25 attributes like budget and revenue.

## Importing libraries 📚

1. __Load__ and __inspect__ the datasets "movies_clean.csv" and "credits.csv". __Identify__ stringified/nested __json columns__ in the __credits__ dataset.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('movies_clean.csv', parse_dates=["release_date"])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44691 entries, 0 to 44690
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     44691 non-null  int64         
 1   title                  44691 non-null  object        
 2   tagline                20284 non-null  object        
 3   release_date           44657 non-null  datetime64[ns]
 4   genres                 42586 non-null  object        
 5   belongs_to_collection  4463 non-null   object        
 6   original_language      44681 non-null  object        
 7   budget_musd            8854 non-null   float64       
 8   revenue_musd           7385 non-null   float64       
 9   production_companies   33356 non-null  object        
 10  production_countries   38835 non-null  object        
 11  vote_count             44691 non-null  float64       
 12  vote_average           42077 non-null  float64       
 13  p

In [4]:
credits = pd.read_csv('credits.csv')
credits

cast  \
0      [{'cast_id': 14, 'character': 'Woody (voice)',...   
1      [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2      [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3      [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4      [{'cast_id': 1, 'character': 'George Banks', '...   
...                                                  ...   
45471  [{'cast_id': 0, 'character': '', 'credit_id': ...   
45472  [{'cast_id': 1002, 'character': 'Sister Angela...   
45473  [{'cast_id': 6, 'character': 'Emily Shaw', 'cr...   
45474  [{'cast_id': 2, 'character': '', 'credit_id': ...   
45475                                                 []   

                                                    crew      id  
0      [{'credit_id': '52fe4284c3a36847f8024f49', 'de...     862  
1      [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...    8844  
2      [{'credit_id': '52fe466a9251416c75077a89', 'de...   15602  
3      [{'credit_id': '52fe44779251416c91011acb', 'de...   31357  
4      [{'credit_id': '52fe44959251416c75039ed7', 'de...   11862  
...                                                  ...     ...  
45471  [{'credit_id': '5894a97d925141426c00818c', 'de...  439050  
45472  [{'credit_id': '52fe4af1c3a36847f81e9b15', 'de...  111109  
45473  [{'credit_id': '52fe4776c3a368484e0c8387', 'de...   67758  
45474  [{'credit_id': '533bccebc3a36844cf0011a7', 'de...  227506  
45475  [{'credit_id': '593e676c92514105b702e68e', 'de...  461257  

[45476 rows x 3 columns]

In [6]:
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45476 entries, 0 to 45475
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    45476 non-null  object
 1   crew    45476 non-null  object
 2   id      45476 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.0+ MB


In [7]:
credits.cast[0]

"[{'cast_id': 14, 'character': 'Woody (voice)', 'credit_id': '52fe4284c3a36847f8024f95', 'gender': 2, 'id': 31, 'name': 'Tom Hanks', 'order': 0, 'profile_path': '/pQFoyx7rp09CJTAb932F2g8Nlho.jpg'}, {'cast_id': 15, 'character': 'Buzz Lightyear (voice)', 'credit_id': '52fe4284c3a36847f8024f99', 'gender': 2, 'id': 12898, 'name': 'Tim Allen', 'order': 1, 'profile_path': '/uX2xVf6pMmPepxnvFWyBtjexzgY.jpg'}, {'cast_id': 16, 'character': 'Mr. Potato Head (voice)', 'credit_id': '52fe4284c3a36847f8024f9d', 'gender': 2, 'id': 7167, 'name': 'Don Rickles', 'order': 2, 'profile_path': '/h5BcaDMPRVLHLDzbQavec4xfSdt.jpg'}, {'cast_id': 17, 'character': 'Slinky Dog (voice)', 'credit_id': '52fe4284c3a36847f8024fa1', 'gender': 2, 'id': 12899, 'name': 'Jim Varney', 'order': 3, 'profile_path': '/eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg'}, {'cast_id': 18, 'character': 'Rex (voice)', 'credit_id': '52fe4284c3a36847f8024fa5', 'gender': 2, 'id': 12900, 'name': 'Wallace Shawn', 'order': 4, 'profile_path': '/oGE6JqPP2xH4t

Problem: Stringified JSON data with double quotes inside, it's not a valid JSON data.

In [8]:
credits.crew[0]

'[{\'credit_id\': \'52fe4284c3a36847f8024f49\', \'department\': \'Directing\', \'gender\': 2, \'id\': 7879, \'job\': \'Director\', \'name\': \'John Lasseter\', \'profile_path\': \'/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f4f\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12891, \'job\': \'Screenplay\', \'name\': \'Joss Whedon\', \'profile_path\': \'/dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f55\', \'department\': \'Writing\', \'gender\': 2, \'id\': 7, \'job\': \'Screenplay\', \'name\': \'Andrew Stanton\', \'profile_path\': \'/pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f5b\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12892, \'job\': \'Screenplay\', \'name\': \'Joel Cohen\', \'profile_path\': \'/dAubAiZcvKFbboWlj7oXOkZnTSu.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f61\', \'department\': \'Writing\', \'gender\': 0, \'id\': 12893, \'job\': \'Screenplay\', \'name\': \'A

## Data exploration:  Preparing the Data for Merge

2. __Drop Duplicates__ in the credits datasets.

In [9]:
credits.head(2)

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew    id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...  8844

In [11]:
credits.id.value_counts()

141971    3
298721    2
9755      2
10991     2
99080     2
         ..
55135     1
15877     1
72272     1
1549      1
461257    1
Name: id, Length: 45432, dtype: int64

id should be unique, but there's duplicates of it.

In [12]:
credits.drop_duplicates(subset='id', inplace = True)

In [13]:
credits.id.value_counts()

862       1
74458     1
296206    1
107308    1
16247     1
         ..
32985     1
44399     1
10138     1
32084     1
461257    1
Name: id, Length: 45432, dtype: int64

After dropping duplicates, now the id is all unique.

## Merging the Data  (Left Join)

3. __Merge/Join__ the datasets movies_clean and credits. -> Add the features __cast__ and __crew__ to the movies_clean dataset.

Download the dataset and prepare a summary of the features available on the dataset including data type (numerical/ categorical), amount of missing data in individual fields. 

In [14]:
# Check whether there are any movies in df for which we don't have cast and crew information in credits.

df[-df.id.isin(credits.id)]

Empty DataFrame
Columns: [id, title, tagline, release_date, genres, belongs_to_collection, original_language, budget_musd, revenue_musd, production_companies, production_countries, vote_count, vote_average, popularity, runtime, overview, spoken_languages, poster_path]
Index: []

In [17]:
#Here, there are 741 movies that are in Credits, not in df

credits[-credits.id.isin(df.id)]

cast  \
189    [{'cast_id': 4, 'character': 'Himself', 'credi...   
554    [{'cast_id': 3, 'character': '', 'credit_id': ...   
682    [{'cast_id': 2, 'character': 'Eunice', 'credit...   
685    [{'cast_id': 0, 'character': 'James Bird', 'cr...   
711                                                   []   
...                                                  ...   
45320  [{'cast_id': 0, 'character': 'Mała metalowa dz...   
45323  [{'cast_id': 0, 'character': 'Krzysiek Buk', '...   
45348                                                 []   
45420  [{'cast_id': 1, 'character': 'Aprel', 'credit_...   
45452                                                 []   

                                                    crew      id  
189                                                   []   56088  
554    [{'credit_id': '52fe4e4dc3a368484e21952d', 'de...  218473  
682    [{'credit_id': '52fe475cc3a36847f81317f7', 'de...   48260  
685    [{'credit_id': '57a321c6c3a3683fa70013d8', 'de...  277270  
711                                                   []  365371  
...                                                  ...     ...  
45320  [{'credit_id': '57ee6832c3a3682a2d003ccd', 'de...  418757  
45323  [{'credit_id': '564fd6a59251414b01005a8a', 'de...  369444  
45348                                                 []  335251  
45420  [{'credit_id': '58d83051c3a368126f0567fe', 'de...  449131  
45452  [{'credit_id': '52fe4684c3a36847f81034f3', 'de...   44330  

[741 rows x 3 columns]

In [18]:
df = df.merge(credits, how = "left", left_on = "id", right_on = "id")
df

id                        title  \
0         862                    Toy Story   
1        8844                      Jumanji   
2       15602             Grumpier Old Men   
3       31357            Waiting to Exhale   
4       11862  Father of the Bride Part II   
...       ...                          ...   
44686  439050                       Subdue   
44687  111109          Century of Birthing   
44688   67758                     Betrayal   
44689  227506             Satan Triumphant   
44690  461257                     Queerama   

                                                 tagline release_date  \
0                                                    NaN   1995-10-30   
1              Roll the dice and unleash the excitement!   1995-12-15   
2      Still Yelling. Still Fighting. Still Ready for...   1995-12-22   
3      Friends are the people who let you be yourself...   1995-12-22   
4      Just When His World Is Back To Normal... He's ...   1995-02-10   
...                                                  ...          ...   
44686         Rising and falling between a man and woman          NaT   
44687                                                NaN   2011-11-17   
44688                             A deadly game of wits.   2003-08-01   
44689                                                NaN   1917-10-21   
44690                                                NaN   2017-06-09   

                         genres           belongs_to_collection  \
0       Animation|Comedy|Family            Toy Story Collection   
1      Adventure|Fantasy|Family                             NaN   
2                Romance|Comedy       Grumpy Old Men Collection   
3          Comedy|Drama|Romance                             NaN   
4                        Comedy  Father of the Bride Collection   
...                         ...                             ...   
44686              Drama|Family                             NaN   
44687                     Drama                             NaN   
44688     Action|Drama|Thriller                             NaN   
44689                       NaN                             NaN   
44690                       NaN                             NaN   

      original_language  budget_musd  revenue_musd  \
0                    en         30.0    373.554033   
1                    en         65.0    262.797249   
2                    en          NaN           NaN   
3                    en         16.0     81.452156   
4                    en          NaN     76.578911   
...                 ...          ...           ...   
44686                fa          NaN           NaN   
44687                tl          NaN           NaN   
44688                en          NaN           NaN   
44689                en          NaN           NaN   
44690                en          NaN           NaN   

                                    production_companies  \
0                                Pixar Animation Studios   
1      TriStar Pictures|Teitler Film|Interscope Commu...   
2                            Warner Bros.|Lancaster Gate   
3                 Twentieth Century Fox Film Corporation   
4              Sandollar Productions|Touchstone Pictures   
...                                                  ...   
44686                                                NaN   
44687                                        Sine Olivia   
44688                            American World Pictures   
44689                                          Yermoliev   
44690                                                NaN   

           production_countries  vote_count  vote_average  popularity  \
0      United States of America      5415.0           7.7   21.946943   
1      United States of America      2413.0           6.9   17.015539   
2      United States of America        92.0           6.5   11.712900   
3      United States of America        34.0           6.1    3.859495   
4      United States of America       173

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44691 entries, 0 to 44690
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     44691 non-null  int64         
 1   title                  44691 non-null  object        
 2   tagline                20284 non-null  object        
 3   release_date           44657 non-null  datetime64[ns]
 4   genres                 42586 non-null  object        
 5   belongs_to_collection  4463 non-null   object        
 6   original_language      44681 non-null  object        
 7   budget_musd            8854 non-null   float64       
 8   revenue_musd           7385 non-null   float64       
 9   production_companies   33356 non-null  object        
 10  production_countries   38835 non-null  object        
 11  vote_count             44691 non-null  float64       
 12  vote_average           42077 non-null  float64       
 13  p

20 Columns, and no missing values now.

## Cleaning and Transforming the new "Cast" Column

4.  __Evaluate__ Python Expressions in the stringified column "cast" and __remove quotes__ ("") where possible.

5. __Determine__ the __cast size__ for all movies (number of actors) and add the additional column "cast_size".

6. __Extract__ all __actor names__ from the column "cast" and __overwrite__ "cast". If a movie has more than one actor, __seperate names by a pipe__ "|".<br>
For example: The value in the first row (Toy Story) should be 'Tom Hanks|Tim Allen|Don Rickles|Jim Varney|Wallace Shawn|John Ratzenberger|Annie Potts|John Morris|Erik von Detten|Laurie Metcalf|R. Lee Ermey|Sarah Freeman|Penn Jillette'.

7. __Inspect__ cast with value_counts(). Do you see anything strange? __Take reasonable measures__!

In [20]:
import ast
import numpy as np

In [21]:
df.cast[0]

"[{'cast_id': 14, 'character': 'Woody (voice)', 'credit_id': '52fe4284c3a36847f8024f95', 'gender': 2, 'id': 31, 'name': 'Tom Hanks', 'order': 0, 'profile_path': '/pQFoyx7rp09CJTAb932F2g8Nlho.jpg'}, {'cast_id': 15, 'character': 'Buzz Lightyear (voice)', 'credit_id': '52fe4284c3a36847f8024f99', 'gender': 2, 'id': 12898, 'name': 'Tim Allen', 'order': 1, 'profile_path': '/uX2xVf6pMmPepxnvFWyBtjexzgY.jpg'}, {'cast_id': 16, 'character': 'Mr. Potato Head (voice)', 'credit_id': '52fe4284c3a36847f8024f9d', 'gender': 2, 'id': 7167, 'name': 'Don Rickles', 'order': 2, 'profile_path': '/h5BcaDMPRVLHLDzbQavec4xfSdt.jpg'}, {'cast_id': 17, 'character': 'Slinky Dog (voice)', 'credit_id': '52fe4284c3a36847f8024fa1', 'gender': 2, 'id': 12899, 'name': 'Jim Varney', 'order': 3, 'profile_path': '/eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg'}, {'cast_id': 18, 'character': 'Rex (voice)', 'credit_id': '52fe4284c3a36847f8024fa5', 'gender': 2, 'id': 12900, 'name': 'Wallace Shawn', 'order': 4, 'profile_path': '/oGE6JqPP2xH4t

Stringified JSON/list.

In [23]:
#apply lambda function to each element

df.cast = df.cast.apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else np.nan)

In [24]:
df.cast[0]

[{'cast_id': 14,
  'character': 'Woody (voice)',
  'credit_id': '52fe4284c3a36847f8024f95',
  'gender': 2,
  'id': 31,
  'name': 'Tom Hanks',
  'order': 0,
  'profile_path': '/pQFoyx7rp09CJTAb932F2g8Nlho.jpg'},
 {'cast_id': 15,
  'character': 'Buzz Lightyear (voice)',
  'credit_id': '52fe4284c3a36847f8024f99',
  'gender': 2,
  'id': 12898,
  'name': 'Tim Allen',
  'order': 1,
  'profile_path': '/uX2xVf6pMmPepxnvFWyBtjexzgY.jpg'},
 {'cast_id': 16,
  'character': 'Mr. Potato Head (voice)',
  'credit_id': '52fe4284c3a36847f8024f9d',
  'gender': 2,
  'id': 7167,
  'name': 'Don Rickles',
  'order': 2,
  'profile_path': '/h5BcaDMPRVLHLDzbQavec4xfSdt.jpg'},
 {'cast_id': 17,
  'character': 'Slinky Dog (voice)',
  'credit_id': '52fe4284c3a36847f8024fa1',
  'gender': 2,
  'id': 12899,
  'name': 'Jim Varney',
  'order': 3,
  'profile_path': '/eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg'},
 {'cast_id': 18,
  'character': 'Rex (voice)',
  'credit_id': '52fe4284c3a36847f8024fa5',
  'gender': 2,
  'id': 12900,
 

Now, it's JSON data. Each element in list is information of actor.

In [25]:
pd.DataFrame(df.cast[0])

cast_id                character                 credit_id  gender  \
0        14            Woody (voice)  52fe4284c3a36847f8024f95       2   
1        15   Buzz Lightyear (voice)  52fe4284c3a36847f8024f99       2   
2        16  Mr. Potato Head (voice)  52fe4284c3a36847f8024f9d       2   
3        17       Slinky Dog (voice)  52fe4284c3a36847f8024fa1       2   
4        18              Rex (voice)  52fe4284c3a36847f8024fa5       2   
5        19             Hamm (voice)  52fe4284c3a36847f8024fa9       2   
6        20          Bo Peep (voice)  52fe4284c3a36847f8024fad       1   
7        26             Andy (voice)  52fe4284c3a36847f8024fc1       0   
8        22              Sid (voice)  52fe4284c3a36847f8024fb1       2   
9        23       Mrs. Davis (voice)  52fe4284c3a36847f8024fb5       1   
10       24         Sergeant (voice)  52fe4284c3a36847f8024fb9       2   
11       25           Hannah (voice)  52fe4284c3a36847f8024fbd       1   
12       27     TV Announcer (voice)  52fe4284c3a36847f8024fc5       2   

         id               name  order                      profile_path  
0        31          Tom Hanks      0  /pQFoyx7rp09CJTAb932F2g8Nlho.jpg  
1     12898          Tim Allen      1  /uX2xVf6pMmPepxnvFWyBtjexzgY.jpg  
2      7167        Don Rickles      2  /h5BcaDMPRVLHLDzbQavec4xfSdt.jpg  
3     12899         Jim Varney      3  /eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg  
4     12900      Wallace Shawn      4  /oGE6JqPP2xH4tNORKNqxbNPYi7u.jpg  
5      7907  John Ratzenberger      5  /yGechiKWL6TJDfVE2KPSJYqdMsY.jpg  
6      8873        Annie Potts      6  /eryXT84RL41jHSJcMy4kS3u9y6w.jpg  
7   1116442        John Morris      7  /vYGyvK4LzeaUCoNSHtsuqJUY15M.jpg  
8     12901    Erik von Detten      8  /twnF1ZaJ1FUNUuo6xLXwcxjayBE.jpg  
9     12133     Laurie Metcalf      9  /unMMIT60eoBM2sN2nyR7EZ2BvvD.jpg  
10     8655       R. Lee Ermey     10  /r8GBqFBjypLUP9VVqDqfZ7wYbSs.jpg  
11    12903      Sarah Freeman     11                              None  
12    37221      Penn Jillette     12  /zmAaXUdx12NRsssgHbk1T31j2x9.jpg

In [27]:
# Creating a new column to determine the cast size.
# Determine with lambda function again.

df["cast_size"] = df.cast.apply(lambda x: len(x))
df.cast_size

0        13
1        26
2         7
3        10
4        12
         ..
44686     3
44687    11
44688    15
44689     5
44690     0
Name: cast_size, Length: 44691, dtype: int64

As you can see, for Toy Story there's a cast size of 13. For the next movie, size of 26 and so on.

In [29]:
df.cast_size.value_counts(dropna=False).head(50)

10    2770
8     2729
7     2710
6     2649
5     2637
9     2557
15    2444
4     2406
11    2283
0     2189
12    2043
13    1799
14    1578
16    1350
1     1302
17    1103
3     1087
2      882
18     811
19     785
20     735
21     591
22     477
23     410
24     354
25     305
26     272
27     246
28     236
31     214
29     194
30     182
32     153
34     137
33     130
35     118
36     106
38      98
37      90
41      87
40      76
42      70
39      70
47      66
43      61
46      59
51      57
45      54
44      54
49      49
Name: cast_size, dtype: int64

The most common cast size is 10.

In [31]:
#Create a new column for casts, separated by |

df.cast = df.cast.apply(lambda x: "|".join(i["name"] for i in x) if isinstance(x, list) else np.nan)
df.cast

0        Tom Hanks|Tim Allen|Don Rickles|Jim Varney|Wal...
1        Robin Williams|Jonathan Hyde|Kirsten Dunst|Bra...
2        Walter Matthau|Jack Lemmon|Ann-Margret|Sophia ...
3        Whitney Houston|Angela Bassett|Loretta Devine|...
4        Steve Martin|Diane Keaton|Martin Short|Kimberl...
                               ...                        
44686              Leila Hatami|Kourosh Tahami|Elham Korda
44687    Angel Aquino|Perry Dizon|Hazel Orencio|Joel To...
44688    Erika Eleniak|Adam Baldwin|Julie du Page|James...
44689    Iwan Mosschuchin|Nathalie Lissenko|Pavel Pavlo...
44690                                                     
Name: cast, Length: 44691, dtype: object

In [32]:
df.cast.value_counts().head(50)

                                                                                                               2189
Georges Méliès                                                                                                   24
Louis Theroux                                                                                                    15
Mel Blanc                                                                                                        12
Jimmy Carr                                                                                                        9
George Carlin                                                                                                     8
Louis C.K.                                                                                                        8
David Attenborough                                                                                                8
Werner Herzog                                                           

The most common value is an empty string with 2189 count.

In [33]:
#Replace the missing values
#nan means Not a Number

df.cast.replace("", np.nan, inplace = True)

In [36]:
df.cast.value_counts().head(5)

Georges Méliès    24
Louis Theroux     15
Mel Blanc         12
Jimmy Carr         9
Werner Herzog      8
Name: cast, dtype: int64

## Cleaning and Transforming the new "Crew" Column

8.  __Evaluate__ Python Expressions in the stringified column "crew" and __remove quotes__ ("") where possible.

9. __Determine__ the __crew size__ for all movies (size of the crew) and add the additional column "crew_size".

10. __Extract__ the __director name__ from the column "crew" and create the new column "director". <br> For example: The value in the first row (Toy Story) should be 'John Lasseter'.

In [35]:
df.crew[0]

'[{\'credit_id\': \'52fe4284c3a36847f8024f49\', \'department\': \'Directing\', \'gender\': 2, \'id\': 7879, \'job\': \'Director\', \'name\': \'John Lasseter\', \'profile_path\': \'/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f4f\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12891, \'job\': \'Screenplay\', \'name\': \'Joss Whedon\', \'profile_path\': \'/dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f55\', \'department\': \'Writing\', \'gender\': 2, \'id\': 7, \'job\': \'Screenplay\', \'name\': \'Andrew Stanton\', \'profile_path\': \'/pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f5b\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12892, \'job\': \'Screenplay\', \'name\': \'Joel Cohen\', \'profile_path\': \'/dAubAiZcvKFbboWlj7oXOkZnTSu.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f61\', \'department\': \'Writing\', \'gender\': 0, \'id\': 12893, \'job\': \'Screenplay\', \'name\': \'A

In [37]:
#ast.literal_eval is to evaluate every string to extract the data

df.crew = df.crew.apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else np.nan)

In [38]:
df.crew[0]

[{'credit_id': '52fe4284c3a36847f8024f49',
  'department': 'Directing',
  'gender': 2,
  'id': 7879,
  'job': 'Director',
  'name': 'John Lasseter',
  'profile_path': '/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f4f',
  'department': 'Writing',
  'gender': 2,
  'id': 12891,
  'job': 'Screenplay',
  'name': 'Joss Whedon',
  'profile_path': '/dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f55',
  'department': 'Writing',
  'gender': 2,
  'id': 7,
  'job': 'Screenplay',
  'name': 'Andrew Stanton',
  'profile_path': '/pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f5b',
  'department': 'Writing',
  'gender': 2,
  'id': 12892,
  'job': 'Screenplay',
  'name': 'Joel Cohen',
  'profile_path': '/dAubAiZcvKFbboWlj7oXOkZnTSu.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f61',
  'department': 'Writing',
  'gender': 0,
  'id': 12893,
  'job': 'Screenplay',
  'name': 'Alec Sokolow',
  'profile_path': '/v79vlRYi94BZUQnkkyzn

In [42]:
pd.DataFrame(df.crew[0])

credit_id  department  gender       id  \
0    52fe4284c3a36847f8024f49   Directing       2     7879   
1    52fe4284c3a36847f8024f4f     Writing       2    12891   
2    52fe4284c3a36847f8024f55     Writing       2        7   
3    52fe4284c3a36847f8024f5b     Writing       2    12892   
4    52fe4284c3a36847f8024f61     Writing       0    12893   
..                        ...         ...     ...      ...   
101  589216c19251412dc2009cb9       Sound       1  1748724   
102  589216ccc3a3680973009274       Sound       0  1748725   
103  589216d79251412dc8009aa0       Sound       0     8067   
104  589216e49251412dcd009a4f  Production       1     7902   
105  589216f39251412dc2009cf3  Production       0    84493   

                        job            name                      profile_path  
0                  Director   John Lasseter  /7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg  
1                Screenplay     Joss Whedon  /dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg  
2                Screenplay  Andrew Stanton  /pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg  
3                Screenplay      Joel Cohen  /dAubAiZcvKFbboWlj7oXOkZnTSu.jpg  
4                Screenplay    Alec Sokolow  /v79vlRYi94BZUQnkkyznbGUZLjT.jpg  
..                      ...             ...                               ...  
101  Assistant Sound Editor   Susan Sanford                              None  
102  Assistant Sound Editor   Susan Popovic                              None  
103  Assistant Sound Editor    Dan Engstrom                              None  
104      Casting Consultant    Ruth Lambert                              None  
105       ADR Voice Casting  Mickie McGowan  /k7TjJBfINsg8vLQxJwos6XObAD6.jpg  

[106 rows x 7 columns]

In [43]:
df['crew_size'] = df.crew.apply(lambda x: len(x))

In [44]:
df.crew_size.value_counts(dropna = False).head(50)

2     6197
3     4951
1     4866
4     3064
5     2240
7     1968
10    1928
6     1867
8     1864
9     1686
11    1555
12    1388
13    1181
14    1017
15     854
16     746
17     633
0      620
18     541
19     420
20     358
21     349
22     287
24     252
23     251
25     220
28     198
26     183
27     174
31     133
30     131
29     125
32     115
34     104
35      97
33      88
36      84
39      83
37      80
40      75
41      69
38      66
44      61
45      56
42      55
46      55
43      52
47      46
48      44
49      42
Name: crew_size, dtype: int64

The most frequent crew size is 2, 3 and 1.

In [46]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i["name"]
    return np.nan

In [47]:
df["director"] = df.crew.apply(get_director)
df.director

0           John Lasseter
1            Joe Johnston
2           Howard Deutch
3         Forest Whitaker
4           Charles Shyer
               ...       
44686    Hamid Nematollah
44687            Lav Diaz
44688      Mark L. Lester
44689    Yakov Protazanov
44690       Daisy Asquith
Name: director, Length: 44691, dtype: object

Directors.

In [48]:
df.director.value_counts(dropna = False).head(50)

NaN                         731
John Ford                    66
Michael Curtiz               65
Werner Herzog                54
Alfred Hitchcock             53
Georges Méliès               49
Woody Allen                  49
Jean-Luc Godard              46
Sidney Lumet                 46
Charlie Chaplin              44
Raoul Walsh                  43
William A. Wellman           42
Henry Hathaway               41
Richard Thorpe               40
Martin Scorsese              39
Ingmar Bergman               39
Fritz Lang                   38
John Huston                  38
George Cukor                 38
Mervyn LeRoy                 37
Robert Altman                37
Robert Wise                  36
Takashi Miike                35
Frank Capra                  35
J. Lee Thompson              35
Clint Eastwood               35
Joseph Sargent               35
Richard Fleischer            35
Roger Corman                 35
Claude Chabrol               35
Howard Hawks                 35
Buster K

The most active/frequent director is John Ford (after the missing values).

## Final Steps

11. __Drop__ the column "crew" and __save__ the dataset in a csv-file.

In [50]:
df.head(2)

id      title                                    tagline release_date  \
0   862  Toy Story                                        NaN   1995-10-30   
1  8844    Jumanji  Roll the dice and unleash the excitement!   1995-12-15   

                     genres belongs_to_collection original_language  \
0   Animation|Comedy|Family  Toy Story Collection                en   
1  Adventure|Fantasy|Family                   NaN                en   

   budget_musd  revenue_musd  \
0         30.0    373.554033   
1         65.0    262.797249   

                                production_companies  ... popularity  runtime  \
0                            Pixar Animation Studios  ...  21.946943     81.0   
1  TriStar Pictures|Teitler Film|Interscope Commu...  ...  17.015539    104.0   

                                            overview  spoken_languages  \
0  Led by Woody, Andy's toys live happily in his ...           English   
1  When siblings Judy and Peter discover an encha...  English|Français   

                                         poster_path  \
0  <img src='http://image.tmdb.org/t/p/w185//rhIR...   
1  <img src='http://image.tmdb.org/t/p/w185//vzmL...   

                                                cast  \
0  Tom Hanks|Tim Allen|Don Rickles|Jim Varney|Wal...   
1  Robin Williams|Jonathan Hyde|Kirsten Dunst|Bra...   

                                                crew cast_size crew_size  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...        13       106   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...        26        16   

        director  
0  John Lasseter  
1   Joe Johnston  

[2 rows x 23 columns]

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44691 entries, 0 to 44690
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     44691 non-null  int64         
 1   title                  44691 non-null  object        
 2   tagline                20284 non-null  object        
 3   release_date           44657 non-null  datetime64[ns]
 4   genres                 42586 non-null  object        
 5   belongs_to_collection  4463 non-null   object        
 6   original_language      44681 non-null  object        
 7   budget_musd            8854 non-null   float64       
 8   revenue_musd           7385 non-null   float64       
 9   production_companies   33356 non-null  object        
 10  production_countries   38835 non-null  object        
 11  vote_count             44691 non-null  float64       
 12  vote_average           42077 non-null  float64       
 13  p

In [52]:
df.drop(columns = ["crew"], inplace = True)

In [55]:
df.to_csv("movies_finish.csv", index = False)

In [59]:
pip install jupyterthemes

Note: you may need to restart the kernel to use updated packages.


In [68]:
import jupyterthemes as jt
from jupyterthemes.stylefx import set_nb_theme
set_nb_theme('monokai')

In [71]:
!jt -t monokai -f fira -fs 10 -nf ptsans -nfs 11 -N -kl -cursw 2 -cursc r -cellw 95% -T